# Programming with Python and a Python Editor (IDE): basic principles of the Python language

Start Pyzo [or IDLE or a similar IDE] or use a ArcGIS Pro Notebook. An Interactive Window (Python Interpreter Window / Python Shell) should be automatically opened. If you are using a Notebook you can directly execute code and results are shown below your "code cell". 
If you use IDEs outside of ArcGIS Pro, make sure that the correct Python shell is loaded/configured (ArcGIS Pro / Python 3). If the opened shell is not adressing Python 3.x you should configure/selecting an additional shell, connecting to your Python 3.x environment (in the case of ArcGIS Pro usually located at C:\Program Files\ArcGIS\Pro\bin\Python\envs\arcgispro-py3\python.exe). The Interactive Window provides you with a way of executing and testing individual lines of code, outputs error messages, and allows you to display output from your code via the print statement. Open a new (empty) document (File > New), where you will write your Python script. Save the empty Script (Save As). 
[If you simply need to write and test a few lines of code, you can use the Interactive Window only] 


 ### Comment

Python and all other programming languages provide a way to document the code that you write. Documentation serves an important purpose in detailing what your script does, who wrote the script, when it was last modified, and generally makes your code easier to understand. Commented lines of code are ignored by the Python interpreter, and serve strictly as documentation. Python uses the "#" sign to comment lines of code. 
The first lines of a script should be commented explanations of the purpose of the script, the author and the date of creation. Add the following lines to your script:



In [ ]:
# Name: Rufai Omowunmi Balogun
# Date: 17th of May, 2021
# Purpose: A basic introduction to Python for GeoProcessing Workflows. 

### Variables

Now we will create some variables. We will use the print statement (explanation in the next lesson) to return the value of the variables:

In [1]:
a = 5
print (a)

5


Some IDE's offer the possibility to check for any syntax errors in your code before running the script. In any case, you will get an error message during execution time if your script contains syntax errors you will e.g. see the following error message:

“Syntax error- invalid syntax” 

In the best case, the cursor will be placed in the line where the error occurs, usually the line number where the error occurs is part of the error message. After correcting the error you can run the script again (e.g. in Pyzo with Ctrl + E).

Run the following lines of code:


In [2]:
b = 5 + 6
print (b)

11


The result of b should now be 11 (sum of 5 + 6). 
Variables can also hold string or list data types:


In [3]:
c = "Hello"
d = ["Python", "Lists", "Variables"] 
print (c)
print  (d[0])
print  (d[2])
print  (d[-1])

Hello
Python
Variables
Variables


Python is case sensitive – try to use now the following code:

In [5]:
strVariable = "Hello"
print (strVariable)

Hello


Now you will get an error message. "NameError: name 'strvariable' is not defined"
= the second variable (strvariable) has not been defined – only 'strVariable' has a value.

### Concatenate strings

String can be concatenated using the (+) operator:

In [6]:
name = "Streets"
type = ".shp"
print (name + type)


Streets.shp


Add the following lines:

In [8]:
i = 1
print (name + str(i) + type)

Streets1.shp


Now you will get an error message similar to “TypeError: must be str, not int”. To add a number to a string expression (concatenate instead of a calculation), the number has to be converted to a string first: Built-in function str() can be used:

In [9]:
print (name + str(i) + type)

Streets1.shp


### Built-in functions

str() is only one of several built-in functions:

Open the Python Manuals (https://docs.python.org/3/ ). 

Click on Library Reference >  Built-in Functions and search for the round function. 

Using the round function:


In [11]:
e = 2.645
f = round (e, 1)
print (f)

2.6


Strings can also be converted to numbers (Caution: this works only, if the string is built of number characters only)

int() function: 


In [12]:
strVariable = "5020"
intVariable = int(strVariable)
print (intVariable *2)

10040


Use the same lines of code but without the int() function:

In [13]:
print (strVariable *2)

50205020


The result is not an error message, but: "50205020". Multiplication of strings is allowed (but not always useful).

### Methods

Methods are like functions, except that they are associated with a particular object.
The usage follows the scheme:	

object.method(parameter) 

Let’s take a mutable list-object. If it is changed, the result has not to be saved in a new variable. Define a list variable and type (in a new line) the variable followed by a “.” – if you are in a ArcGIS Pro Notebook enter the "." and press the "Tabulator" tab. The available methods for lists should be displayed. Select the method append and add a value to the list:



In [15]:
lstVariable = [1,2,3,4,5,6]
lstVariable.append(7)

In [16]:
lstVariable = [1,2,3,4,5,6]
lstVariable.append(7)
print (lstVariable)

[1, 2, 3, 4, 5, 6, 7]


Syntax explanations for such methods can be found in the Python help (in Pyzo called "interactive help" or as always also online).

String also have specific methods: 


In [19]:
strVariable = "c:\\temp \\Data"
print (strVariable.find(":"))
newVariable = strVariable.replace("c:", "d:")
print (newVariable)

1
d:\temp \Data


Caution: The replace method is replacing not only the first occurrence of a character, but also all occurrences in the string!

As you can see, strings are immutable – the methods are not changing the original string. The results of the methods have to be saved in a new variable (in contrast to methods of lists and dictionaries).


### Interactive input:    input function 

The input built-in function is very useful for the testing of scripts (especially outside the ArcGIS environment) and offers the opportunity to interactively request user input during run-time of the script. The input will then be saved within the given variable:

In [20]:
userVariable = input("Select a file name:")
print ("The selected file name is : " + userVariable)

Select a file name:hhh
The selected file name is : hhh


Execute the lines of code and enter any value in the upcoming input-request  (in the Pyzo IDE the input is requested in the interactive window, in some IDEs a pop-up windows aks for input, in Notebooks the input window is shown below the code cell). 
The input will be saved as a string variable. If you want to use a number as interactive user input you have to convert the input values afterwards:

In [21]:
userVariable = input("Please enter a number:")
numCalc = float(userVariable)*float(userVariable)
print (numCalc)

Please enter a number:66
4356.0


=> Your first (very basic) calculation software in Python